<a href="https://colab.research.google.com/github/JFGA96/JFGA96.github.io/blob/master/Proyecto.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##**Para traer el CSV se hace directamente con una cuenta de Google**

In [3]:
!pip install -U -q PyDrive 
  
from pydrive.auth import GoogleAuth 
from pydrive.drive import GoogleDrive 
from google.colab import auth 
from oauth2client.client import GoogleCredentials 
  
  
auth.authenticate_user() 
gauth = GoogleAuth() 
gauth.credentials = GoogleCredentials.get_application_default() 
drive = GoogleDrive(gauth)

## **Se da el link del cual se quiere traer el csv y a traves de pandas se obtiene el dataframe**

In [4]:
link = 'https://drive.google.com/file/d/1uR5s0ol8E1nR_qAMu5iGzwNloRZip8xK/view?usp=s'
  
import pandas as pd 
  
id = link.split("/")[-2] 
  
downloaded = drive.CreateFile({'id':id})  
downloaded.GetContentFile('Reviews.csv')   



In [5]:
  
df = pd.read_csv('Reviews.csv',sep=',') 
df.head(8)

Id   ProductId          UserId                      ProfileName  \
0   1  B001E4KFG0  A3SGXH7AUHU8GW                       delmartian   
1   2  B00813GRG4  A1D87F6ZCVE5NK                           dll pa   
2   3  B000LQOCH0   ABXLMWJIXXAIN  Natalia Corres "Natalia Corres"   
3   4  B000UA0QIQ  A395BORC6FGVXV                             Karl   
4   5  B006K2ZZ7K  A1UQRSCLF8GW1T    Michael D. Bigham "M. Wassir"   
5   6  B006K2ZZ7K   ADT0SRK1MGOEU                   Twoapennything   
6   7  B006K2ZZ7K  A1SP2KVKFXXRU1                David C. Sullivan   
7   8  B006K2ZZ7K  A3JRGQVEQN31IQ               Pamela G. Williams   

   HelpfulnessNumerator  HelpfulnessDenominator  Score        Time  \
0                     1                       1      5  1303862400   
1                     0                       0      1  1346976000   
2                     1                       1      4  1219017600   
3                     3                       3      2  1307923200   
4                     0                       0      5  1350777600   
5                     0                       0      4  1342051200   
6                     0                       0      5  1340150400   
7                     0                       0      5  1336003200   

                                         Summary  \
0                          Good Quality Dog Food   
1                              Not as Advertised   
2                          "Delight" says it all   
3                                 Cough Medicine   
4                                    Great taffy   
5                                     Nice Taffy   
6  Great!  Just as good as the expensive brands!   
7                         Wonderful, tasty taffy   

                                                Text  
0  I have bought several of the Vitality canned d...  
1  Product arrived labeled as Jumbo Salted Peanut...  
2  This is a confection that has been around a fe...  
3  If you are looking for the secret ingredient i...  
4  Great taffy at a great price.  There was a wid...  
5  I got a wild hair for taffy and ordered this f...  
6  This saltwater taffy had great flavors and was...  
7  This taffy is so good.  It is very soft and ch...

In [6]:
cantDatos=df.shape
print(cantDatos)

(568454, 10)


In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 568454 entries, 0 to 568453
Data columns (total 10 columns):
 #   Column                  Non-Null Count   Dtype 
---  ------                  --------------   ----- 
 0   Id                      568454 non-null  int64 
 1   ProductId               568454 non-null  object
 2   UserId                  568454 non-null  object
 3   ProfileName             568438 non-null  object
 4   HelpfulnessNumerator    568454 non-null  int64 
 5   HelpfulnessDenominator  568454 non-null  int64 
 6   Score                   568454 non-null  int64 
 7   Time                    568454 non-null  int64 
 8   Summary                 568427 non-null  object
 9   Text                    568454 non-null  object
dtypes: int64(5), object(5)
memory usage: 43.4+ MB


In [8]:
import matplotlib.pyplot as plt
import seaborn as sb
import numpy as np
%matplotlib inline

# **Verificar las calificaciones dadas para trabajar e identificar la cantidad de comentarios que hay por cada una de las calificaciones**

In [9]:
df['Score'].unique()

array([5, 1, 4, 2, 3])

In [10]:
df['Score'].value_counts()

5    363122
4     80655
1     52268
3     42640
2     29769
Name: Score, dtype: int64

### **Se revisan valores duplicados dentro del dataframe y se mantiene el primero, al ejecutarlo por 'ProductId' se evidencio un cambio en el data frame pero al revisar que muchos de los comentarios estan ligados por el mismo producto se realizo mediante algunas otras variables como: ['UserId','ProductId', 'Profilename', 'Time', 'Score'] Se obtienen los mismos resultados es decir  568454 filas**

In [11]:
df = df.drop_duplicates(df.columns[~df.columns.isin(['UserId','ProductId', 'Profilename', 'Time', 'Score'])],
                        keep='first')


## **Se verifica si existe alguna casilla con datos NaN**


In [12]:
df_check=df.isna()

check_for_any_nan= df.isna().values.any()
# Or
check_for_any_nan= df.isna().any().any()
total_nan_values = df.isna().sum().sum()
print(df_check)
print("NaN Presence:"+str(check_for_any_nan))
print ("Total Number of NaN values:"+str(total_nan_values))

           Id  ProductId  UserId  ProfileName  HelpfulnessNumerator  \
0       False      False   False        False                 False   
1       False      False   False        False                 False   
2       False      False   False        False                 False   
3       False      False   False        False                 False   
4       False      False   False        False                 False   
...       ...        ...     ...          ...                   ...   
568449  False      False   False        False                 False   
568450  False      False   False        False                 False   
568451  False      False   False        False                 False   
568452  False      False   False        False                 False   
568453  False      False   False        False                 False   

        HelpfulnessDenominator  Score   Time  Summary   Text  
0                        False  False  False    False  False  
1                    

## **Se eliminan los valores con casillas nulas. Se encuentran 43 valores y quedan 568411 filas**

In [13]:
df=df.dropna()
print(len(df))

568411


## **Para una mayor relacion al momento de realizar el modelo se deben tomar las muestras de las mismas cantidades**

In [14]:

cuenta = df['Score'].value_counts()
minimo = min(cuenta)
df1=df[df.Score == 5]
df1=df1.sample(minimo)

valor4=df[df.Score == 4]
valor4=valor4.sample(minimo)

valor3=df[df.Score == 3]
valor3=valor3.sample(minimo)

valor2=df[df.Score == 2]
valor2=valor2.sample(minimo)

valor1=df[df.Score == 1]
valor1=valor1.sample(minimo)

df1 = pd.concat([valor1, valor2, valor3, valor4, df1])
df_aleatorio=df1.sample(frac=1).reset_index(drop=True)
#print(df_aleatorio)
#print(df_aleatorio['Score'].value_counts()
print(cuenta)

"""for n in range(1,5):
  valor=df[df.Score == n]
  valor=valor.sample(minimo)
  df1.append(valor)
  print(len(df1))"""


5    363111
4     80655
1     52264
3     42638
2     29743
Name: Score, dtype: int64


'for n in range(1,5):\n  valor=df[df.Score == n]\n  valor=valor.sample(minimo)\n  df1.append(valor)\n  print(len(df1))'

### ** Al momento de revisar la cantidad de datos con Score '2' (que es la cantidad minima de datos) se encuentra que son 29769 filas, este dato es importante ya que si se quiere una estabilidad dentro del modelo lo correcto es tomar la misma cantidad de valores aleatorios de los demas Scores. 
###Se conforma un dataframe con la misma cantidad de datos para cada uno de los Scores con una cantidad de 148715 filas. **

In [15]:
print(df_aleatorio['Score'].value_counts())
print(df_aleatorio.shape)

5    29743
2    29743
3    29743
4    29743
1    29743
Name: Score, dtype: int64
(148715, 10)


## **Para el tratamiento del texto se debe Tokenizar, Eliminar palabras comunes,lemmatizar**

In [16]:
import nltk
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

## **Se utiliza libreria NLTK y se traen las STOPWORDS o "palabras comunes"**

In [17]:
from nltk.corpus import stopwords
import string
punctuation = set(string.punctuation)

stops = set(stopwords.words('english'))
print(len(stops))
wordsFiltered = []


def tokenizacion(sentencia):
    tokens = []
    for token in sentencia.split():
        new_token = []
        for character in token:
            if character not in punctuation:
                new_token.append(character.lower())
        if new_token:
            tokens.append("".join(new_token))
    return tokens

179


In [28]:
stops = set(stopwords.words('english'))

def palabrascomunes(sentencia):
  filtered_word = []
  for token in sentencia.split():
      nuevo_token = []
      for token in stops:
        nuevo_token.append(token)
      if nuevo_token:
          filtered_word.append("".join(nuevo_token))
  return filtered_word


In [ ]:
data = df_aleatorio.head(len(df_aleatorio))["Text"].apply(palabrascomunes)
print(data)
